CenterForOpenScience/osf.io
# [ENG-649] Figshare Article Types Update

*The following is shamelessly copied from @cslzchen's PR text at #9078.  This PR is just that PR with some additional changes*

## Purpose

Our Figshare provider used to only let figshare "fileset"s (article type 4) behave like folders and contain multiple files.  At some point Figshare changed their system so that type 4 ("fileset") is unofficially deprecated and converted to type 3 ("dataset") on creation.

This PR simply allows both filesets and datasets to act as folders without touching the other existing types (except for name changes).

## Changes

### Primary Changes

* Invert the type map so it's now an id-to-type mapping.
* Added a `set` for folder-like type-ids, currently containing 3 (dataset) and 4 (fileset).  Everywhere we used to check for "fileset" or 4, we now check set membership.

### By-products

* Added new types and renamed existing ones.
* Did some minor functionality-preserving code changes and style updates to 1) avoid lines that are too long, to 2) improve if statement by replacing equality check with set membership check and to 3) improve style consistency.

## QA Notes

Please verify that both projects and folders with article type 3 or 4 are listed and can be set as the OSF project storage root during figshare add-on configuration.

Figshare changed their data model and tweaked the API slightly so you will not be able to test this until the [WB side fix](https://openscience.atlassian.net/browse/ENG-380) is released.

In addition, if you are confused on recent weird Figshare behaviors (as I was for quite a while), here is a [Confluence page](https://openscience.atlassian.net/wiki/spaces/SVCS/pages/817037313) that may help.

- Is cross-browser testing required/recommended?
  - **NO** This behavior is browser-independent.
- Is API testing required/recommended?
  - **NO** This change only affects the frontend.
- What pages on the OSF should be tested?
  - The addon configuration page for figshare. Verify successful connection on the Files page
- What edge cases should QA be aware of?
  - The Northern Water Snake looks very similar to the Eastern Copperhead, but is non-venomous and harmless.

## Documentation

Not for the OSF and API.



## Side Effects

None expected.

## Ticket

https://openscience.atlassian.net/browse/ENG-649
